### PR TITLE
Sync with neo LedgerContract

### PR DESCRIPTION
**Pull request by chenzhitong**
Sync with neo https://github.com/neo-project/neo/pull/2215 (LedgerContract )

Incomplete, wait...

### CODE REVIEW

**Comment by chenzhitong** [comment]
**Remark**

```c#
snapshot.Height
↓
NativeContract.Ledger.CurrentIndex(snapshot)

SnapshotView snapshot = Blockchain.Singleton.GetSnapshot();
↓
using var snapshot = Blockchain.Singleton.GetSnapshot();

SnapshotView
↓
SnapshotCache

var currentBlockHash = Snapshot.Blocks[Snapshot.CurrentBlockHash]
var currentBlock = currentBlock = Snapshot.Blocks[Snapshot.CurrentBlockHash];
var previousBlock = Snapshot.Blocks[currentBlock.PrevHash];
↓
var currentBlockHash = NativeContract.Ledger.GetBlockHash(Snapshot, NativeContract.Ledger.CurrentIndex(Snapshot));
var currentBlock = NativeContract.Ledger.GetBlock(Snapshot, currentBlockHash);
var previousBlock = NativeContract.Ledger.GetBlock(Snapshot, currentBlock.PrevHash);

UInt256 hash = Blockchain.Singleton.GetNextBlockHash(block.Hash);
↓
UInt256 hash = NativeContract.Ledger.GetBlockHash(snapshot, block.Index + 1);


Blockchain.Singleton.GetHeader(txState.BlockIndex);
↓
NativeContract.Ledger.GetHeader(snapshot, txState.BlockIndex);
```

**Comment by chenzhitong** [comment]
Is this correct?
```c#
TransactionState txState = Blockchain.Singleton.View.Transactions.TryGet(hash);
var blockIndex = txState.BlockIndex;
↓
TransactionState txState = new TransactionState()
{
    Transaction = tx,
    BlockIndex = NativeContract.Ledger.CurrentIndex(snapshot) - tx.ValidUntilBlock
};
var blockIndex = txState.BlockIndex;
```

**Comment by chenzhitong** [comment]
0 Errors, but RpcServer don't run.
No response on this line.
```c#
using var snapshot = Blockchain.Singleton.GetSnapshot();

or 

var snapshot = Blockchain.Singleton.GetSnapshot().CreateSnapshot();
```

**Comment by roman-khimov** [comment]
Does LevelDB plugin work after 0ded10c9c928daf96e618511597167de11b7c222 fix?

**Comment by superboyiii** [comment]
> Does LevelDB plugin work after [0ded10c](https://github.com/neo-project/neo-modules/commit/0ded10c9c928daf96e618511597167de11b7c222) fix?

Still need more fix on LevelDB. We're working on this.

**Comment by chenzhitong** [comment]
Current Progress
image

**Comment by chenzhitong** [comment]
Current Progress

</img>

File: src/dBFT/ConsensusContext.cs [hunk 1]
```diff
@@ -522,7 +520,7 @@ public void Reset(byte viewNumber)
 
         public void Save()
         {
-            store.PutSync(ConsensusStatePrefix, null, this.ToArray());
+            store.PutSync(null, this.ToArray());
```
Comment (cloud8little) [inline src/dBFT/ConsensusContext.cs:523]: ConsensusStatePrefix should not be removed here.

File: src/dBFT/ConsensusContext.cs [hunk 2]
```diff
@@ -235,7 +236,7 @@ public UInt160 GetSender(int index)
 
         public bool Load()
         {
-            byte[] data = store.TryGet(ConsensusStatePrefix, null);
+            byte[] data = store.TryGet(null);
```
Comment (cloud8little) [inline src/dBFT/ConsensusContext.cs:239]: ConsensusStatePrefix should not be removed here.

**Pull request closed without merging**